packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,225 @@
+import { Tag } from '../cloudformation/tag';
+import { Construct } from './construct';
+import { Token } from './tokens';
+
+/**
+ * ITaggable indicates a entity manages tags via the `tags` property
+ */
+export interface ITaggable {
+    tags: TagManager,
+}
+
+/**
+ * Properties Tags is a dictionary of tags as strings
+ */
+export interface Tags {
+    [key: string]: string
+}
+
+/**
+ * A an object of tags with value and properties
+ *
+ * This is used internally but not exported
+ */
+interface FullTags {
+    [key: string]: {value: string, props?: TagProps};
+}
+
+/**
+ * Properties for a tag
+ */
+export interface TagProps {
+    /**
+     * If true all child taggable `Constructs` will receive this tag
+     *
+     * @default true
+     */
+    propagate?: boolean;
+
+    /**
+     * If set propagated tags from parents will overwrite tags in children
+     *
+     * @default false
+     */
+    propagateOverride?: boolean;
```

Comment:
I have a hard time thinking of use cases where I would want this value to be `false` (as I'm feeling like the "construct assembler" has the most information, so tags applied higher in the tree should basically always propagate and override). Can you enlighten me with some situations where it makes sense for children to set a tag that should NOT be overridden?

The only ones I can think of are tags like "name" or "path". But in those cases, does it not make more sense declare that tag value to be "sticky" or "important" or whatever on the CHILD, rather than require all ancestors in perpetuity to set `propagateOverride=false` when setting tags?

Reply:
@rix0rrr -- so let's first be clear that my viewpoint is heavily influenced by my experience and company and by no means do I consider myself to have full breadth. Tagging is such a customized space that I've actually struggled a lot with the design for 80 % principal.

Cost allocation is the use case in mind for me when designing this feature. A VPC can be shared by multiple teams and commonly cost allocation is done by tags (companies are all over the board with how much to put in 1 account or 1 VPC. My experience is many factors influence here). The VPC itself may be in a top level cost center that amortizes cost across the subsidiaries. While the Dynamo Tables might be divided into specific cost centers. Depending on your team (SRE, AppOps, etc) the VPC and Dynamo may be deployed and operated by a common team. Thus when they create a stack they might default the cost allocation tag to the roll up center, but override each element in the path with a more specific cost center. Another example from my experience is VPC Flow Logs. The cost for flow log and any associated cost of centralization (kinesis streams - normally these are in a central account, but again each company may be different) is actually assigned to a central security business. 

My thinking was at the top level you would put a cost allocation tag, and then at various levels of the tree you would override that tag. If you did nothing the propagation will occur correctly. 

The next line of thinking was that as teams start to implement custom `Construct`s they might want to default these tag values. For example, perhaps you have a requirement that all components are tagged with a `ComplianceLevel` tag that indicates PCI, NIST, SOX, None, etc. The construct developer would default the tag to None, but if the parent was PCI then they would want that propagation to override their value. In this case the assembler would not have to do anything special, at the top level they would set the `ComplianceLevel` and it would propagate everywhere correctly assuming the sub-construct authors configured it correctly. 

I am trying to provide some specific thoughts I have had on the way here, but literally this can be `What if ... ` forever. I'm open to suggestions for the best way evaluate these.